robotoff/taxonomy.py

```diff
@@ -136,16 +145,31 @@ def iter_nodes(self) -> Iterable[TaxonomyNode]:
     def keys(self):
         return self.nodes.keys()
 
-    def find_deepest_nodes(self, nodes: List[TaxonomyNode]) -> List[TaxonomyNode]:
-        """From a list of nodes, find the deepest nodes using the taxonomy."""
+    def find_deepest_nodes(
+        self, nodes: List[TaxonomyNode], ignore_ancestor_gaps=True
+    ) -> List[TaxonomyNode]:
+        """Given a list of nodes, returns the list of nodes where all connected parents
+        within the list have been removed.
+
+        For example, for a taxonomy, 'fish' -> 'salmon' -> 'smoked-salmon':
+
+        ['fish', 'salmon'] -> ['salmon']
+        ['fish', 'smoked-salmon'] -> ['fish', 'smoked-salmon']
+
+        Setting ignore_ancestor_gaps=False, would result in:
+        ['fish', 'smoked-salmon'] -> ['smoked-salmon']
+        """
         excluded: Set[str] = set()
 
-        for node in nodes:
-            for second_node in (
-                n for n in nodes if n.id not in excluded and n.id != node.id
-            ):
-                if node.is_child_of(second_node):
-                    excluded.add(second_node.id)
+        if not ignore_ancestor_gaps:
+            excluded = _ancestor_set(nodes)
+        else:
+            for node in nodes:
+                for second_node in (
+                    n for n in nodes if n.id not in excluded and n.id != node.id
+                ):
+                    if node.is_child_of(second_node):
+                        excluded.add(second_node.id)
```

Comment:
Is this working ? I don't think this will take into account ancestor_gaps, for is_child_of is recursive and not limited to those nodes.

Having unit tests would be useful

Reply:
Sorry, I somehow didn't push all of the changes in this PR - you're correct, since `is_child_of` is recursive we don't need the additional flag. Added the missing changes to the PR.